presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java

```diff
@@ -687,6 +707,53 @@ private static URI findCancelableLeafStage(StageInfo stage)
         return stage.getSelf();
     }
 
+    private boolean planHashesEqual(Optional<CanonicalPlanWithInfo> currentTopLevelPlanHash)
+    {
+        if (currentTopLevelPlanHash.isPresent()) {
+            return currentTopLevelPlanHash.hashCode() == previousQueryTopLevelPlanHash.hashCode();
+        }
+        return false;
+    }
+
+    private boolean retryConditionsMet(QueryResults queryResults)
+    {
+        if (queryResults.getError() == null) {
+            return false;
+        }
+
+        if (!retryQueryWithHistoryBasedOptimizationEnabled(session) && !queryResults.getError().isRetriable()) {
+            return false;
+        }
+
+        // check if we have exceeded the global limit
+        retryCircuitBreaker.incrementFailure();
```

Comment:
This functions is called multiple times above, and it may increment the failure multiple times in one run.

Reply:
@feilong-liu I am looking at the code and am not sure if we need this for the retry